README.md
![markdownd](https://github.com/aerth/markdownd/blob/master/docs/markdownd.png?raw=true)

`markdownd [flags] <directory>`

`markdownd -toc -header theme/header.html -footer theme/footer.html .`

`markdownd -index=gen .`

[![Go Report Card](https://goreportcard.com/badge/github.com/aerth/markdownd)](https://goreportcard.com/report/github.com/aerth/markdownd) 
[![Build Status](https://travis-ci.org/aerth/markdownd.svg?branch=master)](https://travis-ci.org/aerth/markdownd) 

## Markdown Server

  * tries markdown file (.md) in .html request (/index.html tries /index.md first)
  * will serve .html if exists
  * serves static files and downloads if not .html or .md
  * optional indexing (default: off, use -index=gen or -index=README.md)
  * no symlinks
  * no `../` paths
  * raw markdown source requests ( example: `GET /index.md?raw` )
  * custom index page (use flag: `-index README.md`)
  * generates table of contents with `-toc` flag
  * themed html with `-header` and `-footer` flag
  * now with syntax highlighting (use flag: `-syntax`)

## Usage

  * `GET /` will show a 404 unless -index flag is used (-index=gen to generate)
  * `GET /README.md` or `GET /README.html` will process the markdown file and serve HTML.
  * `GET /README.md?raw` will serve raw markdown source
  * To generate index page (with links to files), use `-index=gen`
  * To serve custom `index.md`, use `-index=index.md`

#### Example use case: live preview your git repository's README.md

From your project repository that contains a README.md file, run markdownd like so:

```
markdownd -index=README.md .
```

And visit http://localhost:8080/ in your browser

## Installation

### Compile using Go (from any directory)

```
git clone https://github.com/aerth/markdownd
cd markdownd
make && sudo make install
```

If you don't want to install the server system-wide, or you don't have root privileges, replace last line with:

```
make && make install INSTALLDIR=$HOME/bin
```

### Or using legacy go get

```
GOFLAGS=-tags=netgo,osusergo GOBIN=$HOME/bin go get -v github.com/aerth/markdownd
```

### Download binary for your OS (old versions)

[Latest Release](https://github.com/aerth/markdownd/releases/latest)

Consider installing [go](https://golang.org/dl) and building from source,
Its fast and easy.

## Docker

When using the docker image, markdownd servest the /opt directory,
and exposes port 8080.

You will want to share a directory into /opt and forward the port
(using docker's `-v` and `-p` flags).

For example (modify `$PWD/docs` and `8888` to suit your needs):

`docker run -it -v $PWD/docs:/opt -p 8888:8080 aerth/markdownd`

## Free and Open Source

	The MIT License (MIT)
	
	Copyright (c) 2017-2020  aerth <aerth@riseup.net>
	
	Permission is hereby granted, free of charge, to any person obtaining a 
	copy of this software and associated documentation files (the 
	"Software"), to deal in the Software without restriction, including 
	without limitation the rights to use, copy, modify, merge, publish, 
	distribute, sublicense, and/or sell copies of the Software, and to 
	permit persons to whom the Software is furnished to do so, subject to 
	the following conditions:
	
	The above copyright notice and this permission notice shall be included 
	in all copies or substantial portions of the Software.
	
	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
	OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
	MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
	IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY 
	CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
	TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
	SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
	
## Contributing:

  * pull requests welcome
  * bugs/issues/features very welcome
  * please 'gofmt -w -l -s' before commits


CHANGELOG.md
# changelog

## markdownd 0.0.12
  * generate index file with '-index=gen'
  * minor improvements
  * refactor build scripts
  * use multi-stage docker build (*way* smaller image size)
  * add docker build/run examples

## markdownd 0.0.11
  * use github-flavored-markdown

## markdownd 0.0.10

  * Resolves filepath issue on windows systems (thank you @weaming)
  * Linux and OS X tests are being ran (see travis link on README.md)
  * More accurate response timer
  * Now making sure main ("/") -index file exists, warning if not


main_test.go
package main

import (
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"
)

func TestPrepareDirectory(t *testing.T) {
	dir := "docs"
	dir = prepareDirectory(dir)
	if dir == "docs" {
		t.Log("Wanted absolute pathfrom 'docs', still got 'docs'")
		t.FailNow()
	}

	// give absolute dir
	dir2 := prepareDirectory(dir)
	if dir != dir2 {
		t.Logf("Wanted no change if given absolute path, got: %q != %q", dir, dir2)
		t.FailNow()
	}
}

func TestFileIsGood(t *testing.T) {
	dir := prepareDirectory("docs")
	req := dir + "index.md" // index.md exists
	if !fileisgood(req) {
		// docs is a normal directory, not a symlink. should never be false
		t.Logf("%s should be good, got bad", req)
		t.FailNow()
	}
}

func TestRefuseSymlinks(t *testing.T) {
	dir := prepareDirectory("docs")
	os.Remove(dir + "index.link")
	err := os.Symlink(dir+"index.md", dir+"index.link")
	defer os.Remove(dir + "index.link")
	if err != nil {
		t.Log("Error creating symlink:", err)
		t.FailNow()
	}
	req := dir + "index.link"
	if fileisgood(req) {
		// index.link is a symlink. should never be true
		t.Logf("%s should be good, got bad", req)
		t.FailNow()
	}

}

func TestRefuseDotDots(t *testing.T) {
	dir := prepareDirectory("docs")
	req, _ := http.NewRequest("GET", "/../main.go", nil)
	w := httptest.NewRecorder()
	h := &Handler{
		Root:       http.Dir(dir),
		RootString: dir,
	}
	h.ServeHTTP(w, req)
	resp := w.Result()
	body, _ := ioutil.ReadAll(resp.Body)

	if resp.StatusCode != http.StatusNotFound {
		t.Log("Expected 404, got:", resp.StatusCode)
		t.FailNow()
	}

	if len(body) != 19 {
		t.Log("Expected 404 body, got:", len(body), string(body))
		t.FailNow()
	}

	if string(body) != "404 page not found\n" {
		t.Logf("Expected %q, got: %q", "404 page not found\n", string(body))
		t.FailNow()
	}
}

func TestHTMLHeaderFooter(t *testing.T) {
	dir := prepareDirectory("docs")
	req, _ := http.NewRequest("GET", "/", nil)
	w := httptest.NewRecorder()
	h := &Handler{
		Root:       http.Dir(dir),
		RootString: dir,
		header:     []byte("001"),
		footer:     []byte("002"),
	}
	h.ServeHTTP(w, req)
	resp := w.Result()
	body, _ := ioutil.ReadAll(resp.Body)

	if resp.StatusCode != 200 {
		t.Log("Expected 200, got:", resp.StatusCode)
		t.FailNow()
	}

	bodystr := string(body)

	if !strings.HasPrefix(bodystr, "001") {
		t.Log("Expected prefix of '001'")
		t.Fail()
	}

	if !strings.HasSuffix(bodystr, "002") {
		t.Log("Expected suffix of '002'")
		t.Fail()
	}
}

func sendRequest(req *http.Request) *http.Response {
	dir := prepareDirectory("docs")
	w := httptest.NewRecorder()
	h := &Handler{
		Root:       http.Dir(dir),
		RootString: dir,
	}
	h.ServeHTTP(w, req)
	resp := w.Result()

	return resp

}

func TestBadMethods(t *testing.T) {
	methods := []string{"POST", "PUT", "DELETE", "HEAD",
		"OPTIONS", "TRACE", "CONNECT", "DUMMY"}
	for _, method := range methods {
		req, _ := http.NewRequest(method, "/", nil)
		resp := sendRequest(req)
		if resp.StatusCode != 404 {
			t.Log("Expected 404, got:", resp.StatusCode)
			t.FailNow()
		}

	}
}


build.sh
#!/bin/bash
set -e
GRN="\e[32m"
RED="\e[31m"
RST="\e[0m"

# check for recent go, and grab includes
check_deps(){
	# need go
	if [ -z "$(which go)" ]; then
		echo need latest go, grab from https://golang.org
		exit 111
	fi

	# check go version, warn if out dated
	GOVERSION=$(go version)
	if [  -z "$(echo "$GOVERSION" | grep "go1.8")" ]; then
		printf 'Using unsupported version of go compiler: '
		GOVERSION="$RED""$GOVERSION""$RST"
	else
		GOVERSION="$GRN""$GOVERSION""$RST"
	fi
	printf "$GOVERSION\n"

	# get latest deps if they dont exist
	echo "getting dependencies if they don't exist in "'$GOPATH'
	go get -d -v .
	}


# build static binary

## make building faster (for development only)
# env CGO_ENABLED=0 go install -v


build(){	

	if [ -d "vendor" ]; then
		## ignore vendor dir
		echo moving vendor dir to vendor.mv && echo
		mv -nvi vendor vendor.mv
	fi
	check_deps
	echo building static binary
	time env CGO_ENABLED=0 go build -v -ldflags='-s -w' -o markdownd && echo "built 'markdownd'"
	EXITCODE=$?
	FILESIZE=$(ls -sh markdownd)
	printf "filesize: $FILESIZE\n"

	if [ -x "$(which file)" ]; then
		file markdownd
	fi
	if [ -x "$(which sha256sum)" ]; then
		printf "calculating sha256... "
		sha256sum markdownd
	fi
	if [ -d "vendor.mv" ]; then
		echo putting back vendor dir
		mv -nvi vendor.mv vendor
	fi
	echo exit $EXITCODE
	exit $EXITCODE
}

if [ -z "$@" ]; then 
build && printf "$GRN**** SUCCESS ****$RST\n" && \
exit 0
fi

if [ "package" == "$1" ]; then

set -x
set -e
# make sure we are in go package
if [ -z "$(ls *go)" ]; then
        echo not a go package
        exit 111
fi

# grab make.go and make executable: https://github.com/aerth/make.go/

if [ ! -x "$(which make.go)" ]; then
	echo grab make.go and make executable: https://github.com/aerth/make.go/
	exit 111
fi
env CGO_ENABLED=0 make.go -o pkg/ -all -v
RWD=$(pwd)
cd pkg && sha256sum * > ../SHA256.txt; cd $RWD;

# package each file in 'pkg' dir
for i in $(ls pkg); do
        echo Packaging $i;
        mkdir -p $i;
        cp pkg/$i $i/$i;
        cp README.md $i/;
        cp CHANGELOG.md $i/;
        cp LICENSE $i/;
        cp -avx docs $i/;
        cp -avx theme $i/;
        mkdir -p $i/src; 
        cp *.go $i/src/;
        cp build.sh $i/src/;
        cp SHA256.txt $i/;
        zip -r $i.zip $i/;
        tar czvf $i.tar.gz $i;
        rm -rvf $i;
done
exit 0
fi

echo '
build.sh usage: run with no arguments to build static binary
./build.sh package to build release packages
'


main.go
/*
* The MIT License (MIT)
*
* Copyright (c) 2017  aerth <aerth@riseup.net>
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
 */

// The markdownd command serves markdown, static, and html files.
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/russross/blackfriday"
	"github.com/shurcool/github_flavored_markdown"
	"github.com/sourcegraph/syntaxhighlight"
)

// flags
var (
	addr          = flag.String("http", "127.0.0.1:8080", "address to listen on format 'address:port',\n\tif address is omitted will listen on all interfaces")
	logfile       = flag.String("log", os.Stderr.Name(), "redirect logs to this file")
	indexPage     = flag.String("index", "index.md", "filename to use for paths ending in '/',\n\ttry something like '-index=README.md' or '-index=gen' to generate a simple one.")
	header        = flag.String("header", "", "html header filename for markdown requests")
	footer        = flag.String("footer", "", "html footer filename for markdown requests")
	toc           = flag.Bool("toc", false, "generate table of contents at the top of each markdown page")
	plain         = flag.Bool("plain", false, "disable github flavored markdown")
	syntaxEnabled = flag.Bool("syntax", false, "highlight syntax in .html")
)

// log to file
var logger = log.New(os.Stderr, "[markdownd] ", log.LstdFlags)

const version = "0.0.12"
const sig = "[markdownd v" + version + "] https://github.com/aerth/markdownd"
const serverheader = "markdownd/" + version
const usage = `
USAGE

markdownd [flags] [directory]

EXAMPLES

Serve current directory on 127.0.0.1:8080:
	markdownd .

Serve current directory on all interfaces, port 8080, log to stderr:
	markdownd -log /dev/stderr -http 0.0.0.0:8080 .

Serve 'docs' directory on port 8081, log to 'md.log':
	markdownd -log md.log -http :8081 docs

Serve docs with header, footer, and table of contents. Disable Logs:
	markdownd -log none -header bar.html -footer foo.html -toc docs

Serve docs only on localhost:
	markdownd -http 127.0.0.1:8080 docs
FLAGS
`

// redefine flag Usage
func init() {
	flag.Usage = func() {
		fmt.Println(usage)
		//fmt.Println("FLAGS")
		flag.PrintDefaults()
	}
	rand.Seed(time.Now().UnixNano())
}

// Handler handles markdown requests
type Handler struct {
	Root           http.FileSystem // directory to serve
	RootString     string          // keep directory name for comparing prefix
	header, footer []byte          // for not-raw markdown requests
}

// markdown command
func main() {
	fmt.Println(sig)
	flag.Parse()
	serve(flag.Args())
}

func serve(args []string) {
	// need only 1 argument, the directory to serve
	if len(args) != 1 {
		flag.Usage()
		os.Exit(111)
		return
	}

	// get absolute path of flag.Arg(0)
	dir := flag.Arg(0)
	dir = prepareDirectory(dir)

	if *indexPage != "gen" {
		_, err := os.Stat(dir + *indexPage)
		if err != nil {
			fmt.Fprintf(os.Stderr, "warning: %q not found, did you forget '-index' flag?\n", *indexPage)
		}
	}

	// new markdown handler
	mdhandler := &Handler{
		Root:       http.Dir(dir),
		RootString: dir,
	}

	h := http.DefaultServeMux
	h.Handle("/", mdhandler)
	// print absolute directory we are serving
	println("serving filesystem:", dir)

	// take care of opening log file
	openLogFile()
	println("logging to:", *logfile)

	if *header != "" {
		println("html header:", *header)
		b, err := ioutil.ReadFile(*header)
		if err != nil {
			println(err.Error())
			os.Exit(111)
		}
		mdhandler.header = b
	} else {
		mdhandler.header = []byte("<!DOCTYPE html>\n")
	}

	if *footer != "" {
		println("html footer:", *footer)
		b, err := ioutil.ReadFile(*footer)
		if err != nil {
			println(err.Error())
			os.Exit(111)
		}
		mdhandler.footer = b
	}

	// create a http server
	server := &http.Server{
		Addr:              *addr,
		Handler:           h,
		ErrorLog:          logger,
		MaxHeaderBytes:    (1 << 10), // 1KB
		ReadTimeout:       (time.Second * 5),
		WriteTimeout:      (time.Second * 5),
		ReadHeaderTimeout: (time.Second * 5),
		IdleTimeout:       (time.Second * 5),
	}

	// disable keepalives
	server.SetKeepAlivesEnabled(false)

	// trick to show listening port
	go func() { <-time.After(time.Second); logger.Println("listening:", *addr) }()

	// start serving
	err := server.ListenAndServe()

	// print usage info, probably started wrong or port is occupied
	flag.Usage()

	// always non-nil
	logger.Println(err)

	// any exit is an error
	os.Exit(111)
}

// generate kind-of-unique string
func rfid() string {
	return fmt.Sprintf("request-%04X", rand.Intn(0xFFFF))
}

func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	// all we want is GET
	if r.Method != "GET" {
		logger.Println("bad method:", r.RemoteAddr, r.Method, r.URL.Path, r.UserAgent())
		http.NotFound(w, r)
		return
	}

	// deny requests containing '..'
	if strings.Contains(r.URL.Path, "..") {
		logger.Println("bad path:", r.RemoteAddr, r.Method, r.URL.Path, r.UserAgent())
		http.NotFound(w, r)
		return
	}

	// start timing
	t1 := time.Now()

	// Add Server header
	w.Header().Add("Server", serverheader)

	// Prevent page from being displayed in an iframe
	w.Header().Add("X-Frame-Options", "DENY")

	// generate unique request id
	requestid := rfid()

	// log how long this takes
	defer func(t func() time.Time) {
		logger.Println(requestid, "closed after", t().Sub(t1))
	}(time.Now)

	if *syntaxEnabled && r.URL.Path == "/gh.css" {
		b, err := Asset("static/gh.css")
		if err == nil {
			w.Header().Add("Content-Type", "text/css")
			w.Write(b)
			return
		}
	}

	// abs is not absolute yet
	abs := r.URL.Path[1:] // remove slash prefix
	if abs == "" && *indexPage != "gen" {
		abs = *indexPage
	}

	// '/' suffix, add *index.Page
	if *indexPage != "gen" && strings.HasSuffix(abs, "/") {
		abs += *indexPage
	}

	// still not absolute, prepend root directory to filesrc
	abs = h.RootString + abs

	if *indexPage == "gen" && strings.HasSuffix(r.URL.Path, "/") {
		logger.Println(requestid, "generated index:", abs)
		http.ServeFile(w, r, abs)
		return
	}

	// log now that we have filename
	logger.Println(requestid, r.RemoteAddr, r.Method, r.URL.Path, "->", abs)

	// get absolute path of requested file (could not exist)
	abs, err := filepath.Abs(abs)
	if err != nil {
		logger.Println(requestid, "error resolving absolute path:", err)
		http.NotFound(w, r)
		return
	}

	// .html suffix, but .md exists. choose to serve .md over .html
	if strings.HasSuffix(abs, ".html") {
		trymd := strings.TrimSuffix(abs, ".html") + ".md"
		_, err := os.Open(trymd)
		if err == nil {
			logger.Println(requestid, abs, "->", trymd)
			abs = trymd
		}
	}

	// check if exists, or give 404
	_, err = os.Open(abs)
	if err != nil {
		if strings.Contains(err.Error(), "no such file") {
			logger.Println(requestid, "404", abs)
			http.NotFound(w, r)
			return
		}

		// probably permissions
		logger.Println(requestid, "error opening file:", err, abs)
		http.NotFound(w, r)
		return
	}

	// check if symlink ( to avoid /proc/self/root style attacks )
	if !fileisgood(abs) {
		logger.Printf("%s error: %q is symlink. serving 404", requestid, abs)
		http.NotFound(w, r)
		return
	}

	// compare prefix (alternate way of checking symlink?)
	// above, we checked for abs vs symlink resolved,
	// here lets check if they have the special prefix of "s.Root"
	// probably redundant.
	if !strings.HasPrefix(abs, h.RootString) {
		logger.Println(requestid, "bad path", abs, "doesnt have prefix:", h.RootString)
		http.NotFound(w, r)
		return
	}

	// read bytes (for detecting content type )
	b, err := ioutil.ReadFile(abs)
	if err != nil {
		logger.Printf("%s error reading file: %q", requestid, abs)
		http.NotFound(w, r)
		return
	}

	// detect content type and encoding
	ct := http.DetectContentType(b)

	// serve raw html if exists
	if strings.HasSuffix(abs, ".html") && strings.HasPrefix(ct, "text/html") {
		logger.Println(requestid, "serving raw html:", abs)
		w.Header().Add("Content-Type", "text/html")
		w.Write(b)
		return
	}

	// probably markdown
	if strings.HasSuffix(abs, ".md") && strings.HasPrefix(ct, "text/plain") {
		if strings.Contains(r.URL.RawQuery, "raw") {
			logger.Println(requestid, "raw markdown request:", abs)
			w.Write(b)
			return
		}
		logger.Println(requestid, "serving markdown:", abs)

		md := markdown2html(b)
		if md == nil {
			w.WriteHeader(200)
			return
		}
		w.Header().Add("Content-Type", "text/html")
		w.Write(h.header)
		w.Write(md)
		w.Write(h.footer)
		return
	}

	// fallthrough with http.ServeFile
	logger.Printf("%s serving %s file: %s", requestid, ct, abs)

	http.ServeFile(w, r, abs)
}

// fileisgood returns false if symlink
// comparing absolute vs resolved path is apparently quick and effective
func fileisgood(abs string) bool {

	// sanity check
	if abs == "" {
		return false
	}

	// is absolute really absolute?
	var err error
	if !filepath.IsAbs(abs) {
		abs, err = filepath.Abs(abs)
	}
	if err != nil {
		println(err.Error())
		return false
	}

	// get real path after eval symlinks
	realpath, err := filepath.EvalSymlinks(abs)
	if err != nil {
		println(err.Error())
		return false
	}

	// equality check
	return realpath == abs
}

// prepare root filesystem directory for serving
func prepareDirectory(dir string) string {
	// add slash to dot
	if dir == "." {
		dir += string(os.PathSeparator)
	}

	// become absolute
	var err error
	dir, err = filepath.Abs(dir)
	if err != nil {
		println(err.Error())
		os.Exit(111)
		return err.Error()
	}

	// add trailing slash (for comparing prefix)
	if !strings.HasSuffix(dir, string(os.PathSeparator)) {
		dir += string(os.PathSeparator)
	}

	return dir
}

func markdown2html(in []byte) []byte {
	if len(in) == 0 {
		return nil
	}
	if *plain {
		// default flags
		flags := 0
		if *toc {
			flags |= blackfriday.HTML_TOC
		}
		md := blackfriday.Markdown(
			in, blackfriday.HtmlRenderer(
				// html flags
				flags,
				"", ""),
			// extensions
			0)
		return md
	}

	return github_flavored_markdown.Markdown(in)
}

// use logfile flag and set logger Logger
func openLogFile() {
	switch *logfile {
	case os.Stderr.Name(), "stderr":
		// already stderr
		*logfile = os.Stderr.Name()
	case os.Stdout.Name(), "stdout":
		logger.SetOutput(os.Stdout)
		*logfile = os.Stdout.Name()
	case "none", "no", "null", "/dev/null", "nil", "disabled":
		logger.SetOutput(ioutil.Discard)
		*logfile = os.DevNull
	default:
		func() {
			logger.Printf("Opening log file: %q", *logfile)
			f, err := os.OpenFile(*logfile, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0660)
			if err != nil {
				logger.Fatalf("cant open log file: %s", err)
			}
			logger.SetOutput(f)
		}()
	}

}

func highlightSyntaxHTML(in []byte) (out []byte) {
	out, err := syntaxhighlight.AsHTML(in)
	if err != nil {
		logger.Println("error highlighting syntax:", err)
		return in
	}
	return out
}
